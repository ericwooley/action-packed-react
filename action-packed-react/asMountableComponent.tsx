import * as React from 'react'
export const asMountableComponent = (LoadingComponent: React.ElementType, init: () => Promise<any>) => {
  return class ActionPacedReact extends React.Component {
    jsx: React.ReactElement = <LoadingComponent />
    constructor(props: any) {
      super(props)
      this.loadComponent().catch(e => {
        console.error('Error mounting root component', e)
      })
    }
    loadComponent = async () => {
      this.jsx = await init()
      this.forceUpdate()
    }
    render() {
      return this.jsx
    }
  }
}
