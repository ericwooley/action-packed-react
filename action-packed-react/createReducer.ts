import { Reducer, AnyAction } from 'redux'

export interface IAction<T> extends AnyAction {
  type: string
  payload: T
}

export interface IHandler<T, IState> {
  (state: IState, action: IAction<T>): IState
}

export interface IActionPack<T, IState> {
  (payload: T): IAction<T>
  _handler: IHandler<T, IState>
  _type: string
}

/**
 * Create action pack creates an a action creator which accepts a payload and
 * returns a flux standard action and the action creator has a ._type, for
 * easily checking the type that would be returned, as well as a _handler,
 * which can be used to reduce a state of the expected type.
 * @param type string to be used as the action type.
 * @param handler handler, which should take a state, and standard action,
 *   then return a modified state
 * @see createReducerFromActionPack for combining actionPacks into a single
 *   reducer
 * @example
 * const firstTodo {title: 'Some thing i need to do', done: false}
 * type Todo = typeof firstTodo;
 * const initialTodoState = [firstTodo];
 * type TodoState = typeof initialTodoState;
 * const addTodo = createActionPack<TodoState, Todo>(
 *  'ADD_TODO',
 *  (state, action) => [...state, action.payload]
 *  );
 * // addTodo._type // ADD_TODO
 * const addTodoDemoAction = addTodo({title: 'Something else I need to do', done: true})
 * // {type: 'ADD_TODO', payload: {title: 'Something else I need to do', done: true} }
 * addTodo._handler(initialState, addTodoDemoAction)
 * // [
 * //   {title: 'Some thing i need to do', done: false},
 * //   {title: 'Something else I need to do', done: true}
 * // ]
 */
export const createActionPack = <TState, T>(type: string, handler: IHandler<T, TState>) => {
  const ac: IActionPack<T, TState> & any = (payload: T) => ({
    payload,
    type
  })
  ac._handler = handler
  ac._type = type
  return ac as IActionPack<T, TState>
}

/**
 * Create a reducer with actionCreators as a property, which uses the
 * actionPack handlers to reduce state.
 * @param initialState The initial state to be used by the reducer
 * @param actionPacks Object of action packs whose handlers will be used to
 *   to reduce the the state when the action packs type is dispatched
 * @example
 * const firstTodo {title: 'Some thing i need to do', done: false}
 * type Todo = typeof firstTodo;
 * const initialTodoState = [];
 * const reducer = createReducerFromActionPack(initialTodoState, {
 *   addTodo: createActionPack<TodoState, Todo>(
 *     'ADD_TODO',
 *     (state, action) => [...state, action.payload]
 *   ),
 *   removeTodo: createActionPack<TodoState, Todo>(
 *     'REMOVE_TODO',
 *     (state, action) => state.filter(todo => todo !== action.payload)
 *   ),
 * });
 * let withFirstTodo = reducer(initialState, reducer.actionCreators.addTodo(firstTodo))
 * // [{title: 'Some thing i need to do', done: false}]
 * let withoutFirstTodo = reducer(withFirstTodo, reducer.actionCreators.removeTodo(firstTodo))
 * // []
 *
 */
export const createReducerFromActionPack = <T>(
  initialState: T,
  actionPacks: { [key: string]: IActionPack<any, T> }
): Reducer<T> & { actionCreators: typeof actionPacks } => {
  const handlerMap = Object.values(actionPacks).reduce(
    (hmap, handler) => {
      hmap[handler._type] = handler
      return hmap
    },
    {} as {
      [key: string]: IActionPack<T, any>
    }
  )
  const reducer = (state: T | undefined = initialState, action: AnyAction) => {
    if (handlerMap[action.type]) {
      const updatedState = handlerMap[action.type]._handler(state, action as IAction<any>)
      return updatedState
    }
    return state
  }
  reducer.actionCreators = actionPacks
  return reducer
}
