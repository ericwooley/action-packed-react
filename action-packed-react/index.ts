// export * from './createReducer'
// export * from './link'
// export * from './routeMatcher'
// export * from './RouteMounter'
// export * from './routeReducer'
// export * from './types'

import { createStore, combineReducers, Reducer, Store, applyMiddleware, compose } from 'redux'
import createSagaMiddleware from 'redux-saga'
import { mount } from './RouteMounter'
import {
  routeReducer,
  updateHistory,
  initialState as routeInitialState,
  selectors,
  addUserRoute,
  activateRoute,
  routeCleared
} from './routeReducer'
import { connect, Connect } from 'react-redux'
import {
  ReducerToState,
  IOptions,
  IRoutesMap,
  ReducerObj,
  ICreateRouteOptions,
  IRoutesMapValue,
  IRouteOptions
} from './types'
import { take, select, all, call } from 'redux-saga/effects'
import { createLink } from './link'
import { IRouteComposer, createRouteComposer, IRouteLimitations } from './routeMatcher'

const reducerBase = { _route: routeReducer }
export type BareBonesState = ReducerToState<typeof reducerBase>
export function createApp<R extends { [key: string]: Reducer }>({
  initialState,
  initialReducers,
  history,
  useHashHistory = true,
  render,
  importBaseComponent,
  RouteNotFoundComponent,
  LoadingComponent,
  composeEnhancers = compose
}: IOptions<R>) {
  type IInitialState = BareBonesState & ReducerToState<R>
  let currentReducerObject: typeof reducerBase & R = Object.assign({}, reducerBase, initialReducers)
  const combinedInitialState = Object.assign({}, initialState, {
    _route: routeInitialState
  })
  const routeMap: IRoutesMap = {}
  const sagaMiddleware = createSagaMiddleware()
  const store = createStore(
    combineReducers(currentReducerObject),
    combinedInitialState,
    composeEnhancers(applyMiddleware(sagaMiddleware))
  )
  sagaMiddleware.run(function* rootSaga(): any {
    while (true) {
      yield take(updateHistory._type)
      const currentPath = yield select(selectors.currentPath)
      // const allRoutes = yield select(selectors.routes)
      const outdatedPath = yield select(selectors.activePath)
      if (currentPath !== outdatedPath) {
        // console.log(
        //   'de-activating route',
        //   outdatedPath,
        //   ' to mount',
        //   currentPath
        // )
        const matchingRouteMap: ReturnType<
          typeof selectors.nextPathMatchingRouteMap
        > = yield select(selectors.nextPathMatchingRouteMap)
        const matchingRoutes: ReturnType<typeof selectors.nextPathMatchingRoutes> = yield select(
          selectors.nextPathMatchingRoutes
        )

        const oldMatchingRouteMap: ReturnType<
          typeof selectors.activePathMatchingRouteMap
        > = yield select(selectors.activePathMatchingRouteMap)
        const oldMatchingRoutes: ReturnType<
          typeof selectors.activePathMatchingRoutes
        > = yield select(selectors.activePathMatchingRoutes)
        const noLongerMatchingRoutes = oldMatchingRoutes.filter(r => !matchingRouteMap[r])
        const newlyMatchingRoutes = matchingRoutes.filter(r => !oldMatchingRouteMap[r])
        // console.log('-------> matching routes', {
        //   matchingRoutes,
        //   oldMatchingRoutes,
        //   newlyMatchingRoutes,
        //   noLongerMatchingRoutes,
        //   allRoutes
        // })
        const bundles: {
          pack: IRouteOptions<any, any>
          packLoader: IRoutesMapValue
        }[] = yield all(
          newlyMatchingRoutes
            .filter(route => routeMap[route])
            .map(route => routeMap[route])
            .map(packLoader => async () => {
              packLoader.onRouteMatch()
              const pack = await packLoader.loader()
              return { pack, packLoader }
            })
            .map(call)
        )
        // no more yield here, react can't have another chance to interact with state
        const reducers = [currentReducerObject, ...bundles.map(c => c.pack.reducer)]
        // console.log('----> adding reducer for', newlyMatchingRoutes)
        currentReducerObject = Object.assign({}, ...reducers)
        store.replaceReducer(combineReducers(currentReducerObject))
        store.dispatch(activateRoute(currentPath))
        yield all(
          noLongerMatchingRoutes
            .map(route => async () => {
              const packLoader = routeMap[route]
              if (!packLoader) return
              const pack = await packLoader.loader()
              if (pack.reducer) {
                // console.log('--> removing reducer for', route)
                currentReducerObject = Object.keys(pack.reducer).reduce(
                  (reducer, key) => {
                    reducer[key] = () => null
                    return reducer
                  },
                  { ...(currentReducerObject as any) }
                )
                const reducer = combineReducers(currentReducerObject)
                store.replaceReducer(reducer)
                store.dispatch(routeCleared(route))
                if (pack.onStateCleared) pack.onStateCleared()
              }
            })
            .map(call)
        )
        // console.log('<----------- path switch complete')
      }
    }
  })
  const createSubRoute = <
    IParentReducers extends ReducerObj,
    ParentRouteProps extends IRouteLimitations = {}
  >(
    parentRoute: IRouteComposer<ParentRouteProps>
  ) => <
      ISubReducers extends ReducerObj,
      RouteProps extends Partial<ParentRouteProps>,
      IComponentProps extends { params?: RouteProps; children: React.ReactChildren }
    >(
    route: IRouteComposer<RouteProps> | string,
    {
      reducer
    }: {
      reducer?: () => Promise<ISubReducers>
    },
    options: ICreateRouteOptions = {}
  ) => {
    type IRouteComponent = React.ComponentType<IComponentProps>
    type ILoadRouteComponent = () => Promise<IRouteComponent>
    let component: ILoadRouteComponent = (() => {
      throw new Error('You must set the route component')
    }) as any
    const { onRouteMatch = () => null } = options
    if (typeof route === 'string') route = createRouteComposer<{}>(route)
    const combinedRoute = [parentRoute.route, route.route].join('/')
    type IFullRouteProps = RouteProps & ParentRouteProps
    const routeComposer = createRouteComposer<IFullRouteProps>(combinedRoute)
    routeMap[combinedRoute] = {
      onRouteMatch,
      route: routeComposer,
      parent: routeMap[parentRoute.route],
      loader: async () => {
        const [loadedReducer, loadedComponent] = await Promise.all([
          reducer ? reducer() : Promise.resolve({}),
          component()
        ])
        return {
          component: loadedComponent,
          reducer: loadedReducer
        }
      }
    }
    const setComponent = (componentLoader: ILoadRouteComponent) => {
      component = componentLoader
    }
    store.dispatch(addUserRoute(combinedRoute))
    type ICompleteState = ReducerToState<ISubReducers & IParentReducers>
    const subRoute = {
      Link: createLink(history, routeComposer, useHashHistory),
      routeComposer,
      fullRoute: combinedRoute,
      routeSegment: route,
      setComponent,
      getState: () => (store.getState() as any) as ICompleteState,
      baseSelector: (s: ICompleteState) => s,
      paramSelector: (selectors.params as any) as (s: ICompleteState) => RouteProps,
      connect: <T, OwnProps, H>(
        mapStateToProps: (state: ICompleteState, ownProps: OwnProps) => T,
        handlers?: H
      ) =>
        connect(
          mapStateToProps,
          handlers
        ),
      createSubRoute: createSubRoute<ISubReducers & IParentReducers, IFullRouteProps>(routeComposer)
    }
    return subRoute
  }

  const unlisten = history.listen((location, action) => {
    store.dispatch(
      updateHistory({
        pathname: location.pathname,
        hash: location.hash,
        search: location.search,
        action,
        state: location.state,
        key: location.key
      })
    )
  })
  return {
    shutDown: () => {
      unlisten()
    },
    init: async () => {
      const component = await importBaseComponent
      return new Promise(r => {
        mount(
          {
            pathname: '',
            activeRoute: '',
            routeMap,
            matchingRoutes: [],
            component,
            RouteNotFoundComponent,
            LoadingComponent,
            onMount: r
          },
          store,
          render
        )
        store.dispatch(
          updateHistory({
            pathname: history.location.pathname,
            hash: history.location.hash,
            search: history.location.search,
            action: 'REPLACE',
            state: history.location.state,
            key: history.location.key
          })
        )
      })
    },
    createSubRoute: createSubRoute<IInitialState>(createRouteComposer('')),
    store: store as Store<IInitialState>,
    baseSelector: (s: IInitialState) => s
  }
}
