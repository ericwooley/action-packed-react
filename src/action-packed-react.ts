import {
  createStore,
  combineReducers,
  Reducer,
  Store,
  applyMiddleware
} from 'redux'
import createSagaMiddleware from 'redux-saga'
import { mount } from './RouteMounter'
import {
  routeReducer,
  updateHistory,
  initialState as routeInitialState,
  selectors,
  addUserRoute
} from './routeReducer'
import { connect } from 'react-redux'
import {
  ReducerToState,
  IOptions,
  IRoutesMap,
  ReducerObj,
  IRouteOptionsCreator,
  ICreateRouteOptions,
  IRoutesMapValue,
  IRouteOptions
} from './types'
import { take, select, all, call } from 'redux-saga/effects'

const reducerBase = { _route: routeReducer }
export type BareBonesState = ReducerToState<typeof reducerBase>
export function createApp<R extends { [key: string]: Reducer }>({
  initialState,
  initialReducers,
  history,
  render,
  component
}: IOptions<R>) {
  type IInitialState = BareBonesState & ReducerToState<R>
  let currentReducerObject: typeof reducerBase & R = Object.assign(
    {},
    reducerBase,
    initialReducers
  )
  const combinedInitialState = Object.assign({}, initialState, {
    _route: routeInitialState
  })
  const routeMap: IRoutesMap = {}
  const sagaMiddleware = createSagaMiddleware()
  const store = createStore(
    combineReducers(currentReducerObject),
    combinedInitialState,
    applyMiddleware(sagaMiddleware)
  )
  sagaMiddleware.run(function* rootSaga(): any {
    let oldMatchingRoutes: string[] = []
    let oldMatchingRouteMap: ReturnType<typeof selectors.matchingRouteMap> = {}
    while (true) {
      yield take(updateHistory._type)
      const matchingRouteMap: ReturnType<
        typeof selectors.matchingRouteMap
      > = yield select(selectors.matchingRouteMap)
      const matchingRoutes: ReturnType<
        typeof selectors.matchingRoutes
      > = yield select(selectors.matchingRoutes)
      const noLongerMatchingRoutes = oldMatchingRoutes.filter(
        r => !matchingRouteMap[r]
      )
      const newlyMatchingRoutes = oldMatchingRoutes.filter(
        r => !oldMatchingRouteMap[r]
      )
      oldMatchingRoutes = matchingRoutes
      yield all(
        noLongerMatchingRoutes
          .map(route => async () => {
            const packLoader = routeMap[route]
            if (!packLoader) return
            const pack = await packLoader.loader()
            if (pack.reducer) {
              currentReducerObject = Object.keys(pack.reducer).reduce(
                (reducer, key) => {
                  reducer[key] = () => null
                  return reducer
                },
                { ...(currentReducerObject as any) }
              )
              const reducer = combineReducers(currentReducerObject)
              store.replaceReducer(reducer)
              store.dispatch({
                type: `UNMOUNT_${route}`
              })
            }
          })
          .map(call)
      )
      const bundles: {
        pack: IRouteOptions<any>
        packLoader: IRoutesMapValue
      }[] = yield all(
        newlyMatchingRoutes
          .filter(route => routeMap[route])
          .map(route => routeMap[route])
          .map(packLoader => async () => {
            packLoader.onRouteMatch()
            const pack = await packLoader.loader()
            return { pack, packLoader }
          })
          .map(call)
      )
      const reducers = [
        currentReducerObject,
        ...bundles.map(c => c.pack.reducer)
      ]
      currentReducerObject = Object.assign({}, ...reducers)
      yield call(store.replaceReducer, combineReducers(currentReducerObject))
    }
  })
  const createSubRoute = <IParentState extends ReducerObj>(
    parentRoute: string
  ) => <ISubState extends ReducerObj>(
    route: string,
    routeCreator: IRouteOptionsCreator<ISubState, IParentState>,
    options: ICreateRouteOptions = {}
  ) => {
    const { onRouteMatch = () => null } = options
    const combinedRoute = [parentRoute, route].join('/')
    routeMap[combinedRoute] = {
      onRouteMatch,
      route: route,
      parent: routeMap[parentRoute],
      loader: routeCreator
    }
    store.dispatch(addUserRoute(combinedRoute))
    type CompleteState = ReducerToState<ISubState> & IParentState
    const subRoute = {
      getState: () => (store.getState() as any) as CompleteState,
      baseSelector: (s: CompleteState) => s,
      connect: <T, OwnProps, H>(
        mapStateToProps: (state: CompleteState, ownProps: OwnProps) => T,
        handlers?: H
      ) =>
        connect(
          mapStateToProps,
          handlers
        ),
      createSubRoute: createSubRoute<CompleteState>(combinedRoute)
    }
    return subRoute
  }

  const unlisten = history.listen((location, action) => {
    store.dispatch(
      updateHistory({
        pathname: location.pathname,
        hash: location.hash,
        search: location.search,
        action,
        state: location.state,
        key: location.key
      })
    )
  })
  return {
    shutDown: () => {
      unlisten()
    },
    init: () =>
      new Promise(r => {
        console.log('routes', routeMap)
        mount(
          {
            pathname: '',
            routeMap,
            matchingRoutes: [],
            component
          },
          store,
          render,
          {
            onMount: r
          }
        )
        store.dispatch(
          updateHistory({
            pathname: history.location.pathname,
            hash: history.location.hash,
            search: history.location.search,
            action: 'REPLACE',
            state: history.location.state,
            key: history.location.key
          })
        )
      }),
    createSubRoute: createSubRoute<IInitialState>(''),
    store: store as Store<IInitialState>,
    baseSelector: (s: IInitialState) => s
  }
}
