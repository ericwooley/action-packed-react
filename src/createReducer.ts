let isDevelop = false
try {
  isDevelop = process.env.NODE_ENV === 'development'
} catch (e) {
  // nothing to do here
}
interface IAction<T> {
  type: string
  payload: T
}

interface IHandler<T, IState> {
  (state: IState, action: IAction<T>): IState
}

interface IActionPack<T, IState> {
  (payload: T): { type: string; payload: T }
  _handler: IHandler<T, IState>
  _type: string
}

export const createActionPack = <IState, T>(
  type: string,
  handler: IHandler<T, IState>
) => {
  const ac: IActionPack<T, IState> & any = <T>(payload: T) => ({
    payload,
    type
  })
  ac._handler = handler
  ac._type = type
  return ac as IActionPack<T, IState>
}

export const createReducerFromActionPack = <T>(
  initialState: T,
  handlers: IActionPack<any, T>[],
  { enableWarnings = isDevelop }: { enableWarnings?: boolean } = {}
) => {
  let expectedKeys: { [key: string]: boolean } = {}
  if (enableWarnings) {
    expectedKeys = Object.keys(initialState).reduce(
      (dict, k) => {
        dict[k] = true
        return dict
      },
      {} as { [key: string]: boolean }
    )
  }
  const handlerMap = handlers.reduce(
    (hmap, handler) => {
      hmap[handler._type] = handler
      return hmap
    },
    {} as {
      [key: string]: IActionPack<T, any>
    }
  )
  return (state: T | undefined = initialState, action: IAction<any>) => {
    if (handlerMap[action.type]) {
      const updatedState = handlerMap[action.type]._handler(state, action)
      if (enableWarnings) {
        Object.keys(updatedState).forEach(k => {
          if (!expectedKeys[k]) {
            console.warn(
              'Unexpected state update for key which did not exist on the state before'
            )
          }
        })
      }
      return updatedState
    }
    return state
  }
}
