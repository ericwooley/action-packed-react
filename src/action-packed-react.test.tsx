import * as React from 'react'
import { createApp, BareBonesState } from './action-packed-react'
import { createMemoryHistory } from 'history'
import { mount, render, ReactWrapper } from 'enzyme'
import { Provider } from 'react-redux'
import { createActionPack, createReducerFromActionPack } from './createReducer'
const sanitizeState = (state: BareBonesState) => ({
  ...state,
  _route: {
    ...state._route,
    history: state._route.history.map((item, idx) => ({
      ...item,
      key: `${idx}`
    })),
    key: '<key>'
  }
})
class ProductsComponent extends React.Component<{
  productsLen: number
  test: string
  onClick: () => any
}> {
  render() {
    return (
      <div>
        <h1>ProductsComponent {this.props.productsLen}</h1>
        <p>{this.props.test}</p>
        {this.props.children}
      </div>
    )
  }
}
class ProductSearchComponent extends React.Component<{
  title: string
}> {
  render() {
    return <h1>ProductSearchComponent</h1>
  }
}
type OnMountWrap = (r: ReturnType<typeof mount>) => any
const createBasicApp = (onMountWrap?: OnMountWrap) => {
  const history = createMemoryHistory()
  return {
    history,
    app: createApp({
      history: history,
      initialState: {
        str: '',
        num: 15
      },
      initialReducers: {
        str: () => 'test',
        num: () => 12
      },
      render: jsx => {
        const wrapper = mount(jsx)
        if (onMountWrap) onMountWrap(wrapper)
        return () => null
      }
    })
  }
}

interface Product {
  type: string
}
interface ProductState {
  products: Product[]
}
const productsState: ProductState = {
  products: [{ type: 'test' }]
}
const addProduct = createActionPack<ProductState, { type: string }>(
  'ADD_PRODUCT',
  (state: ProductState, action) => ({
    ...productsState,
    products: [...state.products, action.payload]
  })
)

const productReducer = createReducerFromActionPack(productsState, [addProduct])
const createProducts = (baseApp = createBasicApp()) => {
  const { app, history } = baseApp
  const products = app.createSubRoute('products', async () => ({
    component: ProductsComponent,
    // saga: function* ProductsSaga(): any {},
    reducer: { products: productReducer }
  }))
  return {
    app,
    history,
    products
  }
}
const createProductsSearch = (parentRoute = createProducts()) => {
  const { products, app, history } = parentRoute
  const productSearch = products.createSubRoute('search/:terms', async () => ({
    component: ProductSearchComponent,
    // saga: function* ProductsSearchSaga(): any {},
    reducer: {
      producer: () => null,
      productSearch: (
        state: { test: string },
        action: { type: string; payload: string }
      ) => ({
        ...state,
        test: action.payload
      })
    }
  }))
  return { productSearch, products, app, history }
}
describe('the wooley way fe', () => {
  let baseApp: ReturnType<typeof createBasicApp>
  let wrapper: ReactWrapper
  beforeEach(done => {
    baseApp = createBasicApp(r => {
      wrapper = r
      done()
    })
    baseApp.app.init()
  })
  it('should be setup', () => {
    expect(baseApp.app).toBeTruthy()
    expect(
      baseApp.app.baseSelector(baseApp.app.store.getState())
    ).toMatchSnapshot()
  })
  afterEach(() => baseApp.app.shutDown())
  describe('child route', () => {
    describe('with matching route', () => {
      it('should mount on navigate', done => {
        const productRoute = baseApp.app.createSubRoute(
          'products',
          async () => ({
            component: ProductsComponent,
            // saga: function* ProductsSaga(): any {},
            reducer: {
              products: (
                state: Array<{ type: string }> = [],
                action: { type: string }
              ) => [...state, action]
            }
          }),
          {
            onMount: () => {
              expect(wrapper.html()).toMatchSnapshot('products-full-dom')
              expect(
                render(
                  <Provider store={baseApp.app.store}>
                    <ConnectedProducts test="" />
                  </Provider>
                ).html()
              ).toMatchSnapshot()
              done()
            }
          }
        )
        const ConnectedProducts = productRoute.connect(
          state => ({
            productsLen: state.products.length
          }),
          {
            onClick: () => console.log('click')
          }
        )(ProductsComponent)
        baseApp.app.init()
        baseApp.history.push('/products')
      })
      it('should mount navigate sub-* routes', done => {
        const { app, history, products } = createProducts(baseApp)
        const productRoute = products.createSubRoute(
          'search',
          async () => ({
            component: ProductSearchComponent,
            // saga: function* ProductsSaga(): any {},
            reducer: {
              productsSearch: productReducer
            }
          }),
          {
            onMount: () => {
              expect(wrapper.html()).toMatchSnapshot('products full dom')
              expect(
                render(
                  <Provider store={app.store}>
                    <ConnectedProducts test="" />
                  </Provider>
                ).html()
              ).toMatchSnapshot()
              done()
            }
          }
        )
        const ConnectedProducts = productRoute.connect(
          state => ({
            productsLen: state.productsSearch.products.length
          }),
          {
            onClick: () => console.log('click')
          }
        )(ProductsComponent)
        app.init()
        history.push('/products/search')
      })

      it('should mount/unmount the proper state', async done => {
        const { app, history } = baseApp
        expect(sanitizeState(app.store.getState())).toMatchSnapshot('none')
        await new Promise(r => {
          const productRoute = app.createSubRoute(
            'products',
            async () => ({
              component: ProductsComponent,
              // saga: function* ProductsSaga(): any {},
              reducer: {
                products: productReducer
              }
            }),
            {
              onMount: r,
              onUnMount: () => {
                expect(wrapper.html()).toMatchSnapshot(
                  'full after product unmount'
                )
                expect(sanitizeState(app.store.getState())).toMatchSnapshot(
                  'unmount'
                )
                done()
              }
            }
          )
          app.init()
          history.push('/products')
        })
        expect(sanitizeState(app.store.getState())).toMatchSnapshot('mount')
        history.push('')
      })
      it('should mount/unmount the proper state on subRoutes', async done => {
        const { app, history, products } = createProducts()
        expect(sanitizeState(app.store.getState())).toMatchSnapshot('none')
        await new Promise(r => {
          products.createSubRoute(
            'search',
            async () => ({
              component: ProductSearchComponent,
              // saga: function* ProductsSaga(): any {},
              reducer: {
                productsSearch: productReducer
              }
            }),
            {
              onMount: r,
              onUnMount: () => {
                expect(wrapper.html()).toMatchSnapshot('after unmount')
                expect(sanitizeState(app.store.getState())).toMatchSnapshot(
                  'unmount'
                )
                done()
              }
            }
          )
          app.init()
          history.push('/products/search')
        })
        expect(sanitizeState(app.store.getState())).toMatchSnapshot('mount')
        history.push('')
      })
    })
    describe('without route matching', () => {
      let productsApp: ReturnType<typeof createProducts>
      let productSearchApp: ReturnType<typeof createProductsSearch>
      beforeEach(() => {
        productsApp = createProducts()
        productSearchApp = createProductsSearch()
      })
      it('should create Products', () => {
        expect(productsApp).toBeTruthy()
        expect(productsApp.products.connect).toBeTruthy()
        expect(productsApp.products.createSubRoute).toBeTruthy()
        expect(productsApp.products.getState).toBeTruthy()
        expect(sanitizeState(productsApp.products.getState())).toMatchSnapshot()
        expect(
          sanitizeState(
            productsApp.products.baseSelector(productsApp.products.getState())
          )
        ).toMatchSnapshot()
      })

      it('should connect Products', () => {
        const ConnectedProducts = productsApp.products.connect(
          state => ({
            productsLen: state.products.products.length
          }),
          {
            onClick: () => console.log('click')
          }
        )(ProductsComponent)
        expect(() =>
          render(
            <Provider store={productsApp.app.store}>
              <ConnectedProducts test="" />
            </Provider>
          )
        ).toThrowErrorMatchingInlineSnapshot(
          `"Cannot read property 'products' of undefined"`
        )
      })
      it('should create a sub route', () => {
        // should not error
        const ConnectedProductSearch = productSearchApp.productSearch.connect(
          state => ({
            title: state.productSearch.test
          })
        )(ProductSearchComponent)
        expect(() =>
          render(
            <Provider store={productSearchApp.app.store}>
              <ConnectedProductSearch />
            </Provider>
          )
        ).toThrowErrorMatchingInlineSnapshot(
          `"Cannot read property 'test' of undefined"`
        )
      })
    })
  })
})
