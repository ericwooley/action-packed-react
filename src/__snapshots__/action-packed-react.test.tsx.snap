// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the wooley way fe child route with matching route should mount navigate sub-* routes 1`] = `"<h1>ProductsComponent 1</h1><p></p>"`;

exports[`the wooley way fe child route with matching route should mount navigate sub-* routes 2`] = `"<h1>ProductsComponent 1</h1><p></p>"`;

exports[`the wooley way fe child route with matching route should mount navigate sub-* routes: products full dom 1`] = `"<span>Loading...</span>"`;

exports[`the wooley way fe child route with matching route should mount navigate sub-* routes: products full dom 2`] = `"<div><h1>ProductsComponent </h1><p></p><h1>ProductSearchComponent</h1></div>"`;

exports[`the wooley way fe child route with matching route should mount on navigate 1`] = `"<h1>ProductsComponent 1</h1><p></p>"`;

exports[`the wooley way fe child route with matching route should mount on navigate 2`] = `"<h1>ProductsComponent 2</h1><p></p>"`;

exports[`the wooley way fe child route with matching route should mount on navigate: products-full-dom 1`] = `"<span>Loading...</span>"`;

exports[`the wooley way fe child route with matching route should mount on navigate: products-full-dom 2`] = `"<div><h1>ProductsComponent </h1><p></p></div>"`;

exports[`the wooley way fe child route with matching route should mount/unmount the proper state on subRoutes: after unmount 1`] = `"<span>Loading...</span>"`;

exports[`the wooley way fe child route with matching route should mount/unmount the proper state on subRoutes: mount 1`] = `
Object {
  "_route": Object {
    "action": "PUSH",
    "hash": "",
    "history": Array [
      Object {
        "action": "PUSH",
        "hash": "",
        "key": "0",
        "pathname": "",
        "search": "",
      },
    ],
    "key": "<key>",
    "pathname": "/products/search",
    "search": "",
    "state": undefined,
  },
  "num": 12,
  "products": Object {
    "products": Array [
      Object {
        "type": "test",
      },
    ],
  },
  "productsSearch": Object {
    "products": Array [
      Object {
        "type": "test",
      },
    ],
  },
  "str": "test",
}
`;

exports[`the wooley way fe child route with matching route should mount/unmount the proper state on subRoutes: none 1`] = `
Object {
  "_route": Object {
    "action": "PUSH",
    "hash": "",
    "history": Array [],
    "key": "<key>",
    "pathname": "",
    "search": "",
  },
  "num": 12,
  "str": "test",
}
`;

exports[`the wooley way fe child route with matching route should mount/unmount the proper state on subRoutes: unmount 1`] = `
Object {
  "_route": Object {
    "action": "PUSH",
    "hash": "",
    "history": Array [
      Object {
        "action": "PUSH",
        "hash": "",
        "key": "0",
        "pathname": "",
        "search": "",
      },
      Object {
        "action": "PUSH",
        "hash": "",
        "key": "1",
        "pathname": "/products/search",
        "search": "",
        "state": undefined,
      },
    ],
    "key": "<key>",
    "pathname": "/",
    "search": "",
    "state": undefined,
  },
  "num": 12,
  "products": null,
  "productsSearch": null,
  "str": "test",
}
`;

exports[`the wooley way fe child route with matching route should mount/unmount the proper state: full after product unmount 1`] = `"<div><h1>ProductsComponent </h1><p></p></div>"`;

exports[`the wooley way fe child route with matching route should mount/unmount the proper state: full after product unmount 2`] = `"<div><h1>ProductsComponent </h1><p></p></div>"`;

exports[`the wooley way fe child route with matching route should mount/unmount the proper state: mount 1`] = `
Object {
  "_route": Object {
    "action": "PUSH",
    "hash": "",
    "history": Array [
      Object {
        "action": "PUSH",
        "hash": "",
        "key": "0",
        "pathname": "",
        "search": "",
      },
    ],
    "key": "<key>",
    "pathname": "/products",
    "search": "",
    "state": undefined,
  },
  "num": 12,
  "products": Object {
    "products": Array [
      Object {
        "type": "test",
      },
    ],
  },
  "str": "test",
}
`;

exports[`the wooley way fe child route with matching route should mount/unmount the proper state: none 1`] = `
Object {
  "_route": Object {
    "action": "PUSH",
    "hash": "",
    "history": Array [],
    "key": "<key>",
    "pathname": "",
    "search": "",
  },
  "num": 12,
  "str": "test",
}
`;

exports[`the wooley way fe child route with matching route should mount/unmount the proper state: unmount 1`] = `
Object {
  "_route": Object {
    "action": "PUSH",
    "hash": "",
    "history": Array [
      Object {
        "action": "PUSH",
        "hash": "",
        "key": "0",
        "pathname": "",
        "search": "",
      },
      Object {
        "action": "PUSH",
        "hash": "",
        "key": "1",
        "pathname": "/products",
        "search": "",
        "state": undefined,
      },
    ],
    "key": "<key>",
    "pathname": "/",
    "search": "",
    "state": undefined,
  },
  "num": 12,
  "products": null,
  "str": "test",
}
`;

exports[`the wooley way fe child route with matching route should mount/unmount the proper state: unmount 2`] = `
Object {
  "_route": Object {
    "action": "PUSH",
    "hash": "",
    "history": Array [
      Object {
        "action": "PUSH",
        "hash": "",
        "key": "0",
        "pathname": "",
        "search": "",
      },
      Object {
        "action": "PUSH",
        "hash": "",
        "key": "1",
        "pathname": "/products",
        "search": "",
        "state": undefined,
      },
    ],
    "key": "<key>",
    "pathname": "/",
    "search": "",
    "state": undefined,
  },
  "num": 12,
  "products": null,
  "str": "test",
}
`;

exports[`the wooley way fe child route without route matching should create Products 1`] = `
Object {
  "_route": Object {
    "action": "PUSH",
    "hash": "",
    "history": Array [],
    "key": "<key>",
    "pathname": "",
    "search": "",
  },
  "num": 12,
  "str": "test",
}
`;

exports[`the wooley way fe child route without route matching should create Products 2`] = `
Object {
  "_route": Object {
    "action": "PUSH",
    "hash": "",
    "history": Array [],
    "key": "<key>",
    "pathname": "",
    "search": "",
  },
  "num": 12,
  "str": "test",
}
`;

exports[`the wooley way fe should be setup 1`] = `
Object {
  "_route": Object {
    "action": "PUSH",
    "hash": "",
    "history": Array [],
    "pathname": "",
    "search": "",
  },
  "num": 12,
  "str": "test",
}
`;
